pallets/oracle-feed/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! Oracle pallet to feed values.
+//! Feeding is permissionless given an initial fee for each key.
+//! This pallet do not aggregate/validate any, it just store them by account as
+//! they come. It's expected other pallet read this storage and
+//! aggregate/validate the values.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{fees::PayFee, ValueProvider};
+	use frame_support::{pallet_prelude::*, traits::Time};
+	use frame_system::pallet_prelude::*;
+
+	use crate::weights::WeightInfo;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	type MomentOf<T> = <<T as Config>::Time as Time>::Moment;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Default;
+
+		/// A way to obtain the current time
+		type Time: Time;
+
+		/// Fee for the first time a feeder feeds a value
+		type FirstValuePayFee: PayFee<Self::AccountId>;
+
+		/// The weight information for this pallet extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type FedValues<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		Option<T::AccountId>,
```

Comment:
Should be `Origin` not to store different origins.

Reply:
Actually, `None` corresponds to `Root` and the account to any Signed account. The origin type is not encodable and also contains a variant for unsigned elements that we do not want here.

Seeing that, do you agree as it is?

Reply:
The idea here was `None` for Root and `Some(_)` for a Signed account. But if we want other origins as the bridge origin, we definitely need an origin enum type here.

Reply:
I would just use `T::RuntimeOrigin`,

Reply:
To keep updated this topic for future issues.

`T::RuntimeOrigin` is not encodable and can not be stored, but from `T::RuntimeOrigin`, you can obtain its internal representation (`OriginCaller`) through `OriginTrait` that it is. Although `OriginCaller` has few traits implemented and is less comfortable to use.